### PR TITLE
Add more granular intervals (replace 'days' with generic 'ticks')

### LINKED ISSUE #238: feature: finer granularity options for burndown
While on *most* repos, 1 day is about as granular as you'll need, it would be nice to have the option to set granularity to, say, 0.5 days, or even some number of hours, to look at hackathon projects (which typically span just 24-48 hours).

For such short projects (e.g. https://github.com/launchpals/open-now), the burndowns don't present very interesting data (`granularity=1`):

```
  "Global": [
    [ 343,  0,  0 ],
    [ 309, 470371,  0 ],
    [ 309, 470341, 2073 ],
  ],
```

I'm very interesting in implementing this, though it looks like it won't be all too trivial (a lot of the code seems focused around granularity as days). Any pointers would be appreciated!

it seems like this might be related to #134, but I'm not sure what the plans are for that one, and whether or not it's part of `burndown`

### PR DESCRIPTION
**Pull request by bobheadxi**
closes #238 , more context is in the issue

Seeking feedback! 🙏 probably still needs a bit of work

this PR changes all mentions of "days" to more flexible "ticks", currently configured by `ConfigTicksSinceStartTicksSize` (or `--tick-size`) that accepts an integer value as the number of hours, though tick size is actually tracked as `time.Duration` so even more granular options *could* be exposed (though likely doesn't need to)

`internal/plumbing/day.go` was renamed to `ticks.go` which unfortunately makes the diff a bit hard to review, sorry!

#### default tick size

sets to 24 hours as before, so hopefully it's not significantly breaking

```
❯ hercules --granularity=5 --sampling=5 --burndown https://github.com/launchpals/open-now
hercules:
  version: 9
  hash: 6c50214b59f32b8714469380e1ffcd0cb044517d
  repository: https://github.com/launchpals/open-now
  begin_unix_time: 1548537944
  end_unix_time: 1548645154
  commits: 83
  run_time: 126
Burndown:
  granularity: 5
  sampling: 5
  "project": |-
    472777
```

#### custom tick size

for example, with 1 hour:

```
❯ hercules --granularity=5 --sampling=5 --tick-size=1 --burndown --first-parent https://github.com/launchpals/open-now 
hercules:
  version: 9
  hash: 6c50214b59f32b8714469380e1ffcd0cb044517d
  repository: https://github.com/launchpals/open-now
  begin_unix_time: 1548537944
  end_unix_time: 1548645154
  commits: 10
  run_time: 47
Burndown:
  granularity: 5
  sampling: 5
  "project": |-
    1531        0      0      0      0      0
      1531      0      0      0      0      0
      1258      0    960      0      0      0
      1258      0    960      0      0      0
      1255      0    945      0 468480      0
      1255      0    942      0 468450   2076
```

## bugs

any pointers on these would be appreciated!

* [x] a few failing tests I don't really understand: https://github.com/src-d/hercules/pull/245#issuecomment-473639620
* [x] can't seem to get `labours` to work, but I havent investigated it yet:

*edit* works with `hercules --burndown --first-parent --tick-size=1 --pb https://github.com/bobheadxi/calories | python3 labours.py -f pb -m burndown-project`, I think it's an issue with small repos. that said, the timescales dont work properly with anything other than `--tick-size=24` for now.

```
~/go/src/gopkg.in/src-d/hercules.v9 master* ⇡
❯ hercules --granularity=3 --sampling=3 --tick-size=1 --burndown --pb https://github.com/launchpals/open-now | python3 labours.py -f pb -m burndown-projectdoneing...
project lifetime index: 0.20172258097493653
resampling to year, please wait...
Traceback (most recent call last):
  File "labours.py", line 1942, in <module>
    sys.exit(main())
  File "labours.py", line 1913, in main
    modes[args.mode]()
  File "labours.py", line 1759, in project_burndown
    resample=args.resample))
  File "labours.py", line 627, in load_burndown
    daily[istart:ifinish, (sdt - start).days:].sum(axis=0)
UnboundLocalError: local variable 'sdt' referenced before assignment

~/go/src/gopkg.in/src-d/hercules.v9 master ⇡
❯ hercules --granularity=3 --sampling=3  --burndown --pb https://github.com/launchpals/open-now | python3 labours.py -f pb -m burndown-project 
doneing...
Traceback (most recent call last):
  File "labours.py", line 1942, in <module>
    sys.exit(main())
  File "labours.py", line 1913, in main
    modes[args.mode]()
  File "labours.py", line 1759, in project_burndown
    resample=args.resample))
  File "labours.py", line 589, in load_burndown
    print(name, "lifetime index:", calculate_average_lifetime(matrix))
  File "labours.py", line 439, in calculate_average_lifetime
    lifetimes[i - start] = band[i - 1]
IndexError: index -1 is out of bounds for axis 0 with size 0
```

### CODE REVIEW

**Comment by bobheadxi** [comment]
awkward, really messed up following the DCO bot's suggestion... squashed everything into a single commit and force-pushed 😅

**Comment by bobheadxi** [comment]
@vmarkovtsev , which version of protoc and protoc-gen-go do you use? When I build the proto files I seem to get pretty significant diffs.
                    
libprotoc 3.6.1
protoc-gen-go 1.3.1

**Comment by bobheadxi** [comment]
a few comments on the currently failing tests:

* the pipeline serialization seems to be outputting something different that doesn't seem directly related to the `Days`/`Ticks` change

* for one of the `TicksSinceStart` (formerly `DaysSinceStart`) tests:

```
--- FAIL: TestTicksSinceStartConsumeZero (0.00s)
	ticks_test.go:210: 
			Error Trace:	ticks_test.go:210
			Error:      	Not equal: 
			            	expected: 1970
			            	actual  : 1969
			Test:       	TestTicksSinceStartConsumeZero
```

on my machine, I get `1969`:

```
Running tool: /usr/local/bin/go test -timeout 60s gopkg.in/src-d/hercules.v9/internal/plumbing -run ^(TestTicksSinceStartConsumeZero)$ -v

=== RUN   TestTicksSinceStartConsumeZero
--- PASS: TestTicksSinceStartConsumeZero (0.00s)
PASS
ok  	gopkg.in/src-d/hercules.v9/internal/plumbing	0.109s
Success: Tests passed.
```

File: contrib/_plugin_example/churn_analysis.go [hunk 1]
```diff
@@ -68,14 +68,14 @@ func (churn *ChurnAnalysis) Provides() []string {
 // file_diff - line diff for each commit change
 // changes - list of changed files for each commit
 // blob_cache - set of blobs affected by each commit
-// day - number of days since start for each commit
+// dick - number of ticks since start for each commit
```
Comment (vmarkovtsev) [inline contrib/_plugin_example/churn_analysis.go]: haha lol nice typo

File: internal/plumbing/ticks.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	}
+}
+
+// Configure sets the properties previously published by ListConfigurationOptions().
+func (ticks *TicksSinceStart) Configure(facts map[string]interface{}) error {
+	if val, exists := facts[ConfigTicksSinceStartTickSize].(int); exists {
+		ticks.tickSize = time.Duration(val) * time.Hour
+	} else {
+		// default to 1 day
+		ticks.tickSize = 24 * time.Hour
```
Comment (vmarkovtsev) [inline internal/plumbing/ticks.go]: We should check if `tickSize` is 0 inside `Initialize()` and set it to the default if it is. If there wasn't `ConfigTicksSinceStartTickSize`, we should change the current value to the default only if it equals to 0.

File: internal/plumbing/ticks.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/hercules.v9/internal/core"
+)
+
+// TicksSinceStart provides relative tick information for every commit.
+// It is a PipelineItem.
+type TicksSinceStart struct {
+	core.NoopMerger
+	remote       string
+	tickSize     time.Duration
```
Comment (vmarkovtsev) [inline internal/plumbing/ticks.go]: Let's make it public: `TickSize`

File: internal/plumbing/ticks.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	tickCommits := ticks.commits[tick]
+	if tickCommits == nil {
+		tickCommits = []plumbing.Hash{}
+	}
+
+	exists := false
+	if commit.NumParents() > 0 {
+		for i := range tickCommits {
+			if tickCommits[len(tickCommits)-i-1] == commit.Hash {
+				exists = true
```
Comment (vmarkovtsev) [inline internal/plumbing/ticks.go]: The original code was written a long time ago when I was still learning Go. We should definitely `break` here after setting `exists` to ` true`.

File: leaves/burndown.go [hunk 5]
```diff
@@ -95,11 +95,11 @@ type BurndownAnalysis struct {
...
-	// day is the most recent day index processed.
-	day int
-	// previousDay is the day from the previous sample period -
-	// different from DaysSinceStart.previousDay.
-	previousDay int
+	// tick is the most recent tick index processed.
+	tick int
+	// previousTick is the tick from the previous sample period -
+	// different from TicksSinceStart.previousTick.
+	previousTick int
```
Comment (vmarkovtsev) [inline leaves/burndown.go]: Down to `BurndownResult`: we need to include `tickSize` there.

File: internal/plumbing/ticks.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	if val, exists := facts[ConfigTicksSinceStartTickSize].(int); exists {
+		ticks.tickSize = time.Duration(val) * time.Hour
+	} else {
+		// default to 1 day
+		ticks.tickSize = 24 * time.Hour
+	}
+	if ticks.commits == nil {
+		ticks.commits = map[int][]plumbing.Hash{}
+	}
+	facts[FactCommitsByTick] = ticks.commits
```
Comment (vmarkovtsev) [inline internal/plumbing/ticks.go]: We need to add another fact here which indicates the tick size:
``` go
facts[FactTickSize] = ticks.TickSize
```
Thus `BurndownAnalysis` will be able to read it and save inside `BurndownResult`.

File: contrib/_plugin_example/churn_analysis.go (hunk 1 above)
Comment (bobheadxi) [reply contrib/_plugin_example/churn_analysis.go]: haha oops, fixed in a9b3dfb

File: internal/plumbing/ticks.go (hunk 2 above)
Comment (bobheadxi) [reply internal/plumbing/ticks.go]: I've adjusted this in 1143e52, is that how you want it? a little confused between when to set things in configure, and when to do it on initialize

File: internal/plumbing/ticks.go (hunk 3 above)
Comment (bobheadxi) [reply internal/plumbing/ticks.go]: yep! c0567f8

File: internal/plumbing/ticks.go (hunk 4 above)
Comment (bobheadxi) [reply internal/plumbing/ticks.go]: 👌 b4af94f

File: leaves/burndown.go (hunk 5 above)
Comment (bobheadxi) [reply leaves/burndown.go]: added this in 5b3a96b

File: internal/plumbing/ticks.go (hunk 6 above)
Comment (bobheadxi) [reply internal/plumbing/ticks.go]: also added this in 5b3a96b

**Comment by bobheadxi** [comment]
@vmarkovtsev I realized in all this I missed *actually incorporating* the ticksize into the burndown analysis 😅 so I just had a shot at it in 77cf86a - is there a test I didn't update that I should update / write to verify that this behaves properly for various `tickSize`? Thanks!

Also wondering if you saw my other comment re: my protobuf generation being significantly different from yours, in this comment: https://github.com/src-d/hercules/pull/245#issuecomment-473636661

**Comment by vmarkovtsev** [comment]
Regarding `pb.pb.go`, I use `protoc-gen-gogo`. Refer to `Makefile`.

Regarding 1969 vs 1970, the difference is due to the timezones. I cannot advise what exactly should be changed, but it is definitely because the European timezone is mishandled.

Regarding testing `tickSize`, apart from serialization and deserialization, there should be an error test for merging `BurndownResult` with different tick sizes.

**Comment by bobheadxi** [comment]
> Regarding pb.pb.go, I use protoc-gen-gogo. Refer to Makefile.

Yep, I can generate it correctly using the Makefile - but I think there's either a bit of a *version* mismatch, or I am doing something wrong. I have `protoc-gen-gogo` v1.2.1 installed, and running `make internal/pb/pb.pb.go` I still end up with [strangely large diffs (over 1000 lines)](https://github.com/src-d/hercules/pull/245/files#diff-d433388d7381189954e00660f0c4ab09R46), which seems suspicious. If that's fine I guess it can be ignored 👌 

Just a thought, but with generated code I usually add a CI job to generate code using the expected version of the tools, and then use `git` to exit out if an unexpected diff is detected. Example: https://github.com/ubclaunchpad/pinpoint/blob/master/.travis.yml#L21

**Comment by vmarkovtsev** [comment]
Different tick sizes do not contradict with result merging, but they add complexity to the code which is already complex as hell. We cannot just set the tick size because we need to resample the matrices, and that's not short. So if the tick sizes are different, we should return an error for now.

**Comment by vmarkovtsev** [comment]
I installed `protoc-gen-gogo` about 2 years ago, and I doubt that I have ever updated it since then :smile:. So let's change `Makefile` to work with a specific version and regenerate files with it.

**Comment by bobheadxi** [comment]
> So let's change Makefile to work with a specific version and regenerate files with it.

Probably best to leave this for another PR, if that's okay - we'll want to bump protobuf dependencies, and a bit of scripting will be needed as well (since `go get` plays poorly with versions 😞, so we'll need to cd into directories and fiddle around with things). I'll open a ticket!

File: internal/plumbing/ticks_test.go [hunk 7]
```diff
@@ -203,11 +203,13 @@ func TestTicksSinceStartConsumeZero(t *testing.T) {
 	// depending on where the contributor clones this project from, the remote
 	// reported in the error could either be from gopkg.in or github.com
 	if !strings.Contains(output, "github.com") && !strings.Contains(output, "gopkg.in") {
-		assert.Failf(t, "output should contain either 'github.com' or 'gopkg.in':\n%s", output)
+		assert.Failf(t, "output should contain either 'github.com' or 'gopkg.in'", "got: '%s'", output)
 	}
 	assert.Equal(t, res[DependencyTick].(int), 0)
 	assert.Equal(t, tss.previousTick, 0)
-	assert.Equal(t, tss.tick0.Year(), 1969)
+	if (tss.tick0.Year() != 1969) && (tss.tick0.Year() != 1970) {
```
Comment (vmarkovtsev) [inline internal/plumbing/ticks_test.go]: The root of the problems seems to be that `commit.Committer.When` is initialized to `time.Unix(0, 0)` and that time must be GMT. It is subsequently converted to the local time and we get 1969. Should be fine here.

**Comment by vmarkovtsev** [comment]
Pending tasks:

1. Regenerate `doc/dag.png` for the README header.
2. Update `labours.py` to read the tick size and adjust the X axes.

I can take care of (2) if you wish.

**Comment by bobheadxi** [comment]
1. I've updated the diagram (bc05602)
2. Unfortunately I'm going to be busy this week, so won't be able to look into `labours` for a bit - if you don't mind, can you take care of (2), or leave is as a separate issue and I can take a look at it next week? 🙏 
3. 🚀 

thanks for the reviews and help!

**Pull request merged**